File: dashboard/src/lib/sdk/algorithms.ts
import { sdk } from './litegraph.service';

export type GraphAlgorithmType =
  | 'DegreeCentrality'
  | 'PageRank'
  | 'ClosenessCentrality'
  | 'EigenvectorCentrality'
  | 'BetweennessCentrality'
  | 'WeaklyConnectedComponents'
  | 'StronglyConnectedComponents'
  | 'LabelPropagation'
  | 'Louvain'
  | 'ClusteringCoefficient'
  | 'KCore';

export type GraphExportFormat = 'NodeLinkJson' | 'EdgeList' | 'Graphml';
export type GraphExportAttributeLevel = 'None' | 'Meta' | 'Full';

export type GraphAlgorithmRequest = {
  AlgorithmType: GraphAlgorithmType;
  DampingFactor?: number;
  MaxIterations?: number;
  Tolerance?: number;
  TreatAsUndirected?: boolean;
  MaxResults?: number | null;
  WriteBack?: boolean;
  WriteBackProperty?: string | null;
};

export type GraphAlgorithmNodeResult = {
  NodeGUID: string;
  Name?: string | null;
  Score: number;
  EdgesIn?: number | null;
  EdgesOut?: number | null;
  Community?: number | null;
};

export type GraphAlgorithmResult = {
  Success: boolean;
  TenantGUID?: string;
  GraphGUID?: string;
  AlgorithmType: GraphAlgorithmType;
  NodeCount: number;
  EdgeCount: number;
  Iterations: number;
  Converged: boolean;
  CommunityCount?: number | null;
  ComputeMs: number;
  LoadMs: number;
  WrittenBack: boolean;
  WriteBackProperty?: string | null;
  Nodes: GraphAlgorithmNodeResult[];
};

export type GraphAlgorithmImportRequest = {
  Values: Record<string, Record<string, number>>;
};

export type GraphAlgorithmImportResult = {
  Success: boolean;
  NodesUpdated: number;
};

const getBaseUrl = (): string => {
  const endpoint = sdk.config.endpoint || '/';
  return endpoint.endsWith('/') ? endpoint.slice(0, -1) : endpoint;
};

const buildHeaders = (accept: string, contentType?: string): Record<string, string> => {
  const headers: Record<string, string> = { Accept: accept };
  const defaults = (sdk.config as unknown as { defaultHeaders?: Record<string, string> })
    .defaultHeaders;
  if (defaults) {
    for (const key of Object.keys(defaults)) headers[key] = defaults[key];
  }
  if (contentType) headers['Content-Type'] = contentType;
  return headers;
};

const extractErrorMessage = async (response: Response): Promise<string> => {
  let message = `HTTP ${response.status} ${response.statusText}`;
  try {
    const text = await response.text();
    if (text) {
      try {
        const errorBody = JSON.parse(text);
        message = errorBody?.Description || errorBody?.Message || text;
      } catch {
        message = text;
      }
    }
  } catch {
    // Keep the HTTP status message when the body cannot be read.
  }
  return message;
};

/**
 * Runs a graph algorithm over a single graph.
 * @throws {Error} When the server responds with a non-2xx status.
 */
export const runAlgorithm = async (
  tenantGuid: string,
  graphGuid: string,
  request: GraphAlgorithmRequest
): Promise<GraphAlgorithmResult> => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/graphs/${encodeURIComponent(graphGuid)}/algorithms`;
  const response = await fetch(url, {
    method: 'POST',
    headers: buildHeaders('application/json', 'application/json'),
    body: JSON.stringify(request),
  });
  if (!response.ok) throw new Error(await extractErrorMessage(response));
  return (await response.json()) as GraphAlgorithmResult;
};

/**
 * Exports a graph as a portable projection for external computation (rustworkx / NetworkX).
 * @throws {Error} When the server responds with a non-2xx status.
 */
export const exportGraphProjection = async (
  tenantGuid: string,
  graphGuid: string,
  format: GraphExportFormat = 'NodeLinkJson',
  attributes: GraphExportAttributeLevel = 'Meta'
): Promise<string> => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/graphs/${encodeURIComponent(graphGuid)}/export/projection?format=${encodeURIComponent(
    format
  )}&attributes=${encodeURIComponent(attributes)}`;
  const response = await fetch(url, { method: 'GET', headers: buildHeaders('*/*') });
  if (!response.ok) throw new Error(await extractErrorMessage(response));
  return response.text();
};

/**
 * Imports externally computed per-node values back onto graph nodes.
 * @throws {Error} When the server responds with a non-2xx status.
 */
export const importAlgorithmResults = async (
  tenantGuid: string,
  graphGuid: string,
  request: GraphAlgorithmImportRequest
): Promise<GraphAlgorithmImportResult> => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/graphs/${encodeURIComponent(graphGuid)}/algorithms/import`;
  const response = await fetch(url, {
    method: 'POST',
    headers: buildHeaders('application/json', 'application/json'),
    body: JSON.stringify(request),
  });
  if (!response.ok) throw new Error(await extractErrorMessage(response));
  return (await response.json()) as GraphAlgorithmImportResult;
};

File: sdk/python/src/litegraph_sdk/models/algorithms.py
from typing import Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class GraphAlgorithmNodeResultModel(BaseModel):
    """
    Per-node result produced by a graph algorithm.
    """

    node_guid: Optional[str] = Field(default=None, alias="NodeGUID")
    name: Optional[str] = Field(default=None, alias="Name")
    score: float = Field(default=0.0, alias="Score")
    edges_in: Optional[int] = Field(default=None, alias="EdgesIn")
    edges_out: Optional[int] = Field(default=None, alias="EdgesOut")
    community: Optional[int] = Field(default=None, alias="Community")
    model_config = ConfigDict(populate_by_name=True)


class GraphAlgorithmRequestModel(BaseModel):
    """
    Request to run a graph algorithm over a single graph.
    """

    algorithm_type: str = Field(default="DegreeCentrality", alias="AlgorithmType")
    damping_factor: float = Field(default=0.85, alias="DampingFactor")
    max_iterations: int = Field(default=100, alias="MaxIterations")
    tolerance: float = Field(default=0.000001, alias="Tolerance")
    treat_as_undirected: bool = Field(default=False, alias="TreatAsUndirected")
    max_results: Optional[int] = Field(default=None, alias="MaxResults")
    write_back: bool = Field(default=False, alias="WriteBack")
    write_back_property: Optional[str] = Field(default=None, alias="WriteBackProperty")
    use_cache: bool = Field(default=False, alias="UseCache")
    model_config = ConfigDict(populate_by_name=True)


class GraphAlgorithmResultModel(BaseModel):
    """
    Result of a graph algorithm run.
    """

    success: bool = Field(default=True, alias="Success")
    tenant_guid: Optional[str] = Field(default=None, alias="TenantGUID")
    graph_guid: Optional[str] = Field(default=None, alias="GraphGUID")
    algorithm_type: str = Field(default="DegreeCentrality", alias="AlgorithmType")
    node_count: int = Field(default=0, alias="NodeCount")
    edge_count: int = Field(default=0, alias="EdgeCount")
    iterations: int = Field(default=0, alias="Iterations")
    converged: bool = Field(default=True, alias="Converged")
    community_count: Optional[int] = Field(default=None, alias="CommunityCount")
    compute_ms: float = Field(default=0.0, alias="ComputeMs")
    load_ms: float = Field(default=0.0, alias="LoadMs")
    from_cache: bool = Field(default=False, alias="FromCache")
    written_back: bool = Field(default=False, alias="WrittenBack")
    write_back_property: Optional[str] = Field(default=None, alias="WriteBackProperty")
    nodes: List[GraphAlgorithmNodeResultModel] = Field(default_factory=list, alias="Nodes")
    model_config = ConfigDict(populate_by_name=True)


class GenerateEmbeddingsRequestModel(BaseModel):
    """
    Request to generate node embeddings using the tenant's active embedding endpoint.
    """

    max_nodes: Optional[int] = Field(default=None, alias="MaxNodes")
    skip_nodes_with_vectors: bool = Field(default=True, alias="SkipNodesWithVectors")
    model_config = ConfigDict(populate_by_name=True)


class GenerateEmbeddingsResultModel(BaseModel):
    """
    Result of a node embedding generation run.
    """

    success: bool = Field(default=True, alias="Success")
    nodes_embedded: int = Field(default=0, alias="NodesEmbedded")
    nodes_skipped: int = Field(default=0, alias="NodesSkipped")
    model: Optional[str] = Field(default=None, alias="Model")
    dimensionality: int = Field(default=0, alias="Dimensionality")
    endpoint_guid: Optional[str] = Field(default=None, alias="EndpointGUID")
    model_config = ConfigDict(populate_by_name=True)


class GraphAlgorithmImportRequestModel(BaseModel):
    """
    Request to import externally computed per-node values back onto graph nodes.
    Values maps node GUID strings to a dictionary of property name to numeric value.
    """

    values: Dict[str, Dict[str, float]] = Field(default_factory=dict, alias="Values")
    model_config = ConfigDict(populate_by_name=True)
